### PR TITLE
[ECP-8465] If capture fields for open-invoice payment methods can be combined into 1 drop-down

### PR DESCRIPTION
**Pull request by khushboo-singhvi**
**Description**
After Investigating if capture fields for open-invoice payment methods can be combined into 1 drop-down, finally it was merged in one dropdown with the dropdown values as Immediate, Manual, On Shipment

**Tested scenarios**
- Successful test for all 3 scenarios.

### CODE REVIEW

File: Helper/Config.php [hunk 1]
```diff
@@ -519,12 +519,14 @@ public function debugLogsEnabled($storeId): bool
 
     public function getAutoCaptureOpenInvoice(int $storeId): bool
     {
-        return $this->getConfigData('auto_capture_openinvoice', self::XML_ADYEN_ABSTRACT_PREFIX, $storeId, true);
+        $captureForOpenInvoice = $this->getConfigData('capture_for_openinvoice', self::XML_ADYEN_ABSTRACT_PREFIX, $storeId, true);
+        if($captureForOpenInvoice == 'auto') return true;
```
Comment (candemiralp) [inline Helper/Config.php:523]: Could you please use `strcmp()` for the comparison and assign the string `auto` to a constant?

File: Helper/Config.php [hunk 2]
```diff
@@ -559,25 +557,12 @@ public function getConfigurationMode(int $storeId): string
...
 
-    /**
-     * Returns global configuration values
-     *
-     * @param $field
-     * @param null|int|string $storeId
-     * @return mixed
-     */
-    public function getAdyenAbstractConfigData($field, $storeId = null)
+    public function getAdyenAbstractConfigData($field, int|string $storeId = null): mixed
```
Comment (candemiralp) [inline Helper/Config.php:560]: Type of the `storeId` argument should be `int` and `field` argument might be string.

File: Helper/Config.php [hunk 3]
```diff
@@ -559,25 +557,12 @@ public function getConfigurationMode(int $storeId): string
...
     }
 
-    /**
-     * Retrieve the Live endpoint prefix key
-     *
-     * @param null|int|string $storeId
-     * @return string
-     */
-    public function getLiveEndpointPrefix($storeId = null)
+    public function getLiveEndpointPrefix(int|string $storeId = null): ?string
```
Comment (candemiralp) [inline Helper/Config.php:565]: Type of the storeId should be int.

File: Helper/Config.php [hunk 4]
```diff
@@ -588,106 +573,52 @@ public function getLiveEndpointPrefix($storeId = null)
...
 
-    /**
-     * Returns adyen_oneclick configuraiton values as flag.
-     *
-     * @param $field
-     * @param $storeId
-     * @return bool|mixed
-     */
-    public function getAdyenOneclickConfigDataFlag($field, $storeId = null)
+    public function getAdyenOneclickConfigDataFlag($field, int|string $storeId = null): mixed
```
Comment (candemiralp) [inline Helper/Config.php:606]: Argument type declarations please.

File: Helper/Config.php [hunk 5]
```diff
@@ -588,106 +573,52 @@ public function getLiveEndpointPrefix($storeId = null)
...
     {
         return $this->getConfigData($field, 'adyen_oneclick', $storeId, true);
     }
 
-    public function isPerStoreBillingAgreement($storeId) //Only use of Flag above
+    public function isPerStoreBillingAgreement($storeId): bool //Only use of Flag above
```
Comment (candemiralp) [inline Helper/Config.php:611]: Argument type declarations please

File: Helper/Config.php [hunk 6]
```diff
@@ -588,106 +573,52 @@ public function getLiveEndpointPrefix($storeId = null)
...
-
-    /**
-     * Gives back adyen_boleto configuration values
-     *
-     * @param $field
-     * @param null|int|string $storeId
-     * @return mixed
-     */
-    public function getAdyenBoletoConfigData($field, $storeId = null)
+    public function getAdyenBoletoConfigData($field, int|string $storeId = null): mixed
```
Comment (candemiralp) [inline Helper/Config.php:616]: Argument type declarations please

File: Helper/Config.php [hunk 7]
```diff
@@ -588,106 +573,52 @@ public function getLiveEndpointPrefix($storeId = null)
...
     }
 
-    /**
-     * Retrieve the Checkout frontend region
-     *
-     * @param null|int|string $storeId
-     * @return string
-     */
-    public function getCheckoutFrontendRegion($storeId = null)
+    public function getCheckoutFrontendRegion(int|string $storeId = null): ?string
```
Comment (candemiralp) [inline Helper/Config.php:621]: Argument type declarations please

File: Helper/Config.php [hunk 8]
```diff
@@ -703,16 +634,7 @@ public function getRatePayId($storeId = null)
...
-     * Retrieve information from payment configuration
-     *
-     * @param string $field
-     * @param string $xmlPrefix
-     * @param int $storeId
-     * @param bool|false $flag
-     * @return bool|mixed
-     */
-    public function getConfigData($field, $xmlPrefix, $storeId, $flag = false)
+    public function getConfigData($field, $xmlPrefix, $storeId, $flag = false): mixed
```
Comment (candemiralp) [inline Helper/Config.php:637]: Argument type declarations please

File: Helper/Config.php [hunk 9]
```diff
@@ -723,7 +645,7 @@ public function getConfigData($field, $xmlPrefix, $storeId, $flag = false)
         }
     }
 
-    public function setConfigData($value, $field, $xmlPrefix, $scope = ScopeConfigInterface::SCOPE_TYPE_DEFAULT)
+    public function setConfigData($value, $field, $xmlPrefix, $scope = ScopeConfigInterface::SCOPE_TYPE_DEFAULT): void
```
Comment (candemiralp) [inline Helper/Config.php:648]: Argument type declarations please

File: Helper/Data.php [hunk 10]
```diff
@@ -309,6 +322,15 @@ public function getCaptureModes()
         ];
     }
 
+    public function getOpenInvoiceCaptureModes()
+    {
+        return [
+            'auto' => 'immediate',
+            'manual' => 'manual',
+            'onshipment' => 'on shipment'
```
Comment (candemiralp) [inline Helper/Data.php:330]:
```suggestion
            'auto' => 'Immediate',
            'manual' => 'Manual',
            'onshipment' => 'On shipment'
```

File: Helper/SupportFormHelper.php [hunk 11]
```diff
@@ -386,8 +386,8 @@ public function getConfigData(): array
             'boletoDeliveryDays' => $boletoDeliveryDays,
             'isPayByLinkEnabled' => $isPayByLinkEnabled ? 'Yes' : 'No',
             'useManualCaptureForPaypal' => $useManualCaptureForPaypal ? 'Yes' : 'No',
-            'captureOnShipmentForOpenInvoice' => $captureOnShipmentForOpenInvoice ? 'Yes' : 'No',
-            'autoCaptureOpenInvoice' => $autoCaptureOpenInvoice ? 'Yes' : 'No',
+            'captureOnShipmentForOpenInvoice' => $captureOnShipmentForOpenInvoice,
+            'autoCaptureOpenInvoice' => $autoCaptureOpenInvoice,
```
Comment (candemiralp) [inline Helper/SupportFormHelper.php:390]: As a follow-up to previous comment, those values will be identical. You can merge them into one field. After changing the values in this function, you need to update support form templates [here](https://github.com/Adyen/adyen-magento2/tree/develop/view/adminhtml/email) which use those fields.

File: Observer/BeforeShipmentObserver.php [hunk 12]
```diff
@@ -80,13 +80,13 @@ public function execute(Observer $observer)
...
-        $captureOnShipment = $this->configHelper->getConfigData(
-            'capture_on_shipment',
+        $openInvoiceCapture = $this->configHelper->getConfigData(
+            'capture_for_openinvoice',
             self::XML_ADYEN_ABSTRACT_PREFIX,
             $order->getStoreId()
         );
 
-        if (!$captureOnShipment) {
+        if ($openInvoiceCapture != 'onshipment') {
```
Comment (candemiralp) [inline Observer/BeforeShipmentObserver.php:89]: Can you please use `strcmp` and move the string to constants preferably the same place with the first comment.

File: etc/adminhtml/system/adyen_capture_settings.xml [hunk 13]
```diff
@@ -24,17 +24,11 @@
...
-            <source_model>Magento\Config\Model\Config\Source\Yesno</source_model>
-            <config_path>payment/adyen_abstract/auto_capture_openinvoice</config_path>
+        <field id="capture_for_openinvoice" translate="label" type="select" sortOrder="20" showInDefault="1" showInWebsite="1" showInStore="1">
+            <label>Capture for OpenInvoice Methods</label>
+            <tooltip>Applicable for Klarna, Afterpay Touch, Ratepay, FacilyPay/Oney, Affirm, Clearpay, Zip and PayBright. By default, Open Invoice methods are set to manual capture. If you want auto capture you need to contact magento@adyen.com. After approval has been given you can set this option to 'Immediate'. If you want to do a capture request to Adyen if you do a shipment then set it to 'On Shipment'.</tooltip>
+            <source_model>Adyen\Payment\Model\Config\Source\OpenInvoiceCaptureMode</source_model>
+            <config_path>payment/adyen_abstract/capture_for_openinvoice</config_path>
```
Comment (candemiralp) [inline etc/adminhtml/system/adyen_capture_settings.xml:31]: Could you please revisit label and tooltip together with our UX writer? Maybe we can simplify this text.

**Comment by candemiralp** [comment]
Could you please check the code smells also?

**Review by candemiralp — APPROVED** [review]

File: Helper/Config.php [hunk 14]
```diff
@@ -434,12 +435,18 @@ public function debugLogsEnabled($storeId): bool
...
     public function getAutoCaptureOpenInvoice(int $storeId): bool
     {
-        return $this->getConfigData('auto_capture_openinvoice', self::XML_ADYEN_ABSTRACT_PREFIX, $storeId, true);
+        $captureForOpenInvoice = $this->getConfigData('capture_for_openinvoice', self::XML_ADYEN_ABSTRACT_PREFIX, $storeId, true);
+        if (strcmp((string) $captureForOpenInvoice, self::AUTO_CAPTURE_OPENINVOICE) === 0) {
+            return true;
+        }
+        else {
+            return false;
+        }
```
Comment (hossam-adyen) [inline Helper/Config.php:444]:
```suggestion
        return strcmp((string) $captureForOpenInvoice, self::AUTO_CAPTURE_OPENINVOICE) === 0
```

Comment (khushboo-singhvi) [reply Helper/Config.php:444]: As per my findings strcmp returns the value. 
ref: https://www.php.net/manual/en/function.strcmp.php

Comment (hossam-adyen) [reply Helper/Config.php:444]: My comment is related to the `if` statement, you can remove it as my code suggestion stated

**Review by hossam-adyen — APPROVED** [review]

**Review by candemiralp — APPROVED** [review]

**Pull request merged**